internal/querycoordv2/meta/resource_manager.go

```diff
@@ -602,25 +623,75 @@ func (rm *ResourceManager) selectMissingRecoverSourceRG(rg *ResourceGroup) *Reso
 
 // recoverRedundantNodeRG recover resource group by transfer node to other resource group.
 func (rm *ResourceManager) recoverRedundantNodeRG(rgName string) error {
+	rgFilter := func(targetRG *ResourceGroup) bool {
+		sourceRG := rm.groups[rgName]
+		for _, node := range sourceRG.GetNodes() {
+			nodeInfo := rm.nodeMgr.Get(node)
+			if nodeInfo != nil && targetRG.AcceptNode(nodeInfo) {
+				return true
+			}
+		}
+		return false
+	}
+
 	for rm.groups[rgName].RedundantNumOfNodes() > 0 {
-		rg := rm.groups[rgName]
-		targetRG := rm.selectRedundantRecoverTargetRG(rg)
+		sourceRG := rm.groups[rgName]
+		targetRG := rm.selectRedundantRecoverTargetRG(sourceRG, rgFilter)
 		if targetRG == nil {
 			log.Info("failed to select redundant recover target resource group, please check resource group configuration if as expected.",
-				zap.String("rgName", rg.GetName()))
+				zap.String("rgName", sourceRG.GetName()))
 			return errors.New("all resource group reach limits")
 		}
 
-		nodeID, err := rm.transferOneNodeFromRGToRG(rg, targetRG)
+		nodeID, err := rm.transferOneNodeFromRGToRG(sourceRG, targetRG)
 		if err != nil {
 			log.Warn("failed to recover redundant node by transfer node to other resource group",
-				zap.String("sourceRG", rg.GetName()),
+				zap.String("sourceRG", sourceRG.GetName()),
 				zap.String("targetRG", targetRG.GetName()),
 				zap.Error(err))
 			return err
 		}
 		log.Info("recover redundant node by transfer node to other resource group",
-			zap.String("sourceRG", rg.GetName()),
+			zap.String("sourceRG", sourceRG.GetName()),
+			zap.String("targetRG", targetRG.GetName()),
+			zap.Int64("nodeID", nodeID),
+		)
+	}
+	return nil
+}
+
+// recoverDirtyNodeRG recover resource group by move out all node which doesn't meet node label requirement.
+func (rm *ResourceManager) recoverDirtyNodeRG(rgName string) error {
+	rgFilter := func(targetRG *ResourceGroup) bool {
+		sourceRG := rm.groups[rgName]
+		for _, node := range sourceRG.GetNodes() {
+			nodeInfo := rm.nodeMgr.Get(node)
+			if nodeInfo != nil && targetRG.AcceptNode(nodeInfo) {
+				return true
+			}
+		}
+		return false
+	}
+
+	for len(rm.groups[rgName].DirtyNodes(rm.nodeMgr)) > 0 {
+		sourceRG := rm.groups[rgName]
+		targetRG := rm.selectRedundantRecoverTargetRG(sourceRG, rgFilter)
```

Comment:
`DefaultResourceGroup` is a fallback rg that always receive the redundant node.
I suggest that make the `DefaultResourceGroup` to be a resource group that cannot be modified at the `resource group filter`. 
Otherwise, `infinite-loop` recovery may be occurred here.